### DISCUSSION TITLE
中文讨论须知

### DISCUSSION THREAD

**Discussion by unknwon** [flamego/flamego#87] [中文讨论]
感谢您使用 Flamego 并愿意参与讨论，您可以在该子板块使用中文进行常规讨论，但该板块不接受问题反馈和功能提需。

如需进行问题反馈和功能提需，请使用我们预设的[问题模板](https://github.com/flamego/flamego/issues/new/choose)并使用英文完成提交。

可加入 “Flamego 社区交流群” QQ 群 910083715。

**Comment by chenkaiC4** [comment]
嗨，无闻。很高兴看到你和小伙伴们创建 flamego 这个项目。

在中大型项目中，DI 能够给业务逻辑编写，带来很高的效率提升。

我曾经在go社区找各种 DI 的解决方案，但没有谁是从框架底层去兼容的。

后来索性就使用 nodejs 社区的 nestjs 框架了，一款基于DI和装饰器的模块化框架。

我想知道，对于 flamego 在 v1.0 后，进一步的发展计划是什么？

**Reply by unknwon** [reply]
Hello @chenkaiC4，目前 flamego 核心库已经稳定，所以就是完善文档和添加更多的官方中间件。Go 1.18 范型出来可能会尝试看是否有机会通过范型的机制对函数调用进行性能优化。

**Comment by aohanhongzhi** [comment]
我貌似没有在官网文档中找到flamesgo关于ORM层的解决方案，这个是要借助现有成熟的项目[GORM](https://gorm.io/)来解决吗？

**Reply by unknwon** [reply]
对的，Flamego 只负责到路由层，建议配合 GORM 使用

**Comment by liaohongxing** [comment]
1. 返回值带有 err 情况下 , return_handler 支持对 err 进行定制, 最好能再套一个flamego.Handler。根据不同的err返回不同的结果。根据不同客户端返回不同相应，这也是go社区对错误处理常见的范式。当然这是可配置，没有配置Handler时还是按原来的处理。
2. c.Param  c.Query  等系列方法支持默认值, 比如获取分页  c.QueryInt("page_size", "15")  , 这是很常见需求。
3. 如何配置内置的*log.Logger的 flags  output，例如让其显示短行号Lshortfile

**Reply by unknwon** [reply]
感谢反馈！

> 1. 返回值带有 err 情况下 , return_handler 支持对 err 进行定制, 最好能再套一个flamego.Handler。根据不同的err返回不同的结果。根据不同客户端返回不同相应，这也是go社区对错误处理常见的范式。当然这是可配置，没有配置Handler时还是按原来的处理。

这个目前就是支持的，https://github.com/flamego/flamego/blob/main/return_handler.go 因为目前实现不太优雅所以没有文档细致描述。目前正在更的视频教程第 6 期就会讲你所提到的类似场景的例子：https://github.com/unknwon/flamego-basics



> 2\. c.Param  c.Query  等系列方法支持默认值, 比如获取分页  c.QueryInt("page_size", "15")  , 这是很常见需求。

这个可以有，也不破坏目前 API 结构

> 3. 如何配置内置的*log.Logger的 flags  output，例如让其显示短行号Lshortfile

可以通过在 Flame 实例上直接[重载服务](https://flamego.cn/custom-services.html#%E9%87%8D%E8%BD%BD%E6%9C%8D%E5%8A%A1)实现，视频教程第 5 期也讲到了。

**Reply by unknwon** [reply]
第二点我开了个 issue https://github.com/flamego/flamego/issues/125

**Comment by byene0923** [comment]
flamesgo 有没有计划增加对FileServer的支持？结合1.16的embed和fs。

**Reply by unknwon** [reply]
你是指这个吗？https://flamego.cn/core-services.html#%E7%A4%BA%E4%BE%8B-%E5%93%8D%E5%BA%94-embed-fs

**Reply by byene0923** [reply]
是的，已经实现了啊！，这个是可以配合prefix一起使用的嘛？

**Reply by unknwon** [reply]
可以配合的

**Comment by sg4395** [comment]
你好 flamego server是0.0.0.0：2803 但我要通过什么access 这个port呢 localhost吗 还是就是browser？我试了几次都说page not available 或者no data is sent， 有办法把上传的信息直接save + 导出到文档吗？

**Reply by unknwon** [reply]
你好，我不是很理解你想要问的点，我建议可以先看下 https://flamego.cn/starter-guide.html

**Reply by sg4395** [reply]
你好，我是刚接触编程，就是我现在server开始后 不知道去什么窗口来看client传给server的数据。terminal就显示 [Flamego] Listening on 0.0.0.0:2830 (development) ，我是想问我怎么到这个2830， 是通过browser 输入吗 还是 browser上面输入localhost：2830？麻烦你了

server 截图
<img width="980" alt="Screen Shot 2022-12-08 at 11 30 02 PM" src="https://user-images.githubusercontent.com/80510852/206488316-f10b577d-b038-41bd-af39-0e05b4b28502.png">

<img width="518" alt="Screen Shot 2022-12-08 at 11 35 25 PM" src="https://user-images.githubusercontent.com/80510852/206488531-8be9ea19-d486-4de9-9af3-4ce1f8dcf724.png">
<img width="318" alt="Screen Shot 2022-12-08 at 11 35 39 PM" src="https://user-images.githubusercontent.com/80510852/206488547-d9533775-27cc-48c9-9d54-a7a59ee3e722.png">

**Reply by sg4395** [reply]
<img width="482" alt="Screen Shot 2022-12-08 at 11 37 06 PM" src="https://user-images.githubusercontent.com/80510852/206489440-6de6ac9e-b829-4849-9fa9-05b6c9ae91ca.png">

**Reply by unknwon** [reply]
了解，那你先看下我发的链接，就基本知道怎么才能不返回404了

**Comment by aohanhongzhi** [comment]
flamego的日志怎么由 [logrus](github.com/sirupsen/logrus)接管啊？
![image](https://github.com/flamego/flamego/assets/37319319/99c29435-f58c-414a-9b3c-68b6ff58e2cb)

**Reply by aohanhongzhi** [reply]
```go
import (
  log "github.com/sirupsen/logrus"
  charmbraceletLog "github.com/charmbracelet/log"
)

func main(){
	f := flamego.New()
	f.Use(
		Logger(),
		flamego.Renderer(
		flamego.RenderOptions{
			JSONIndent: "  ",
		},
	))
}

func Logger() flamego.Handler {
	return flamego.LoggerInvoker(func(ctx flamego.Context, logger *charmbraceletLog.Logger) {
		started := time.Now()

		log.Print("Started",
			"method", ctx.Request().Method,
			"path", ctx.Request().RequestURI,
			"remote", ctx.RemoteAddr(),
			)

		w := ctx.ResponseWriter()
		ctx.Next()

		log.Print("Completed",
			"method", ctx.Request().Method,
			"path", ctx.Request().RequestURI,
			"status", w.Status(),
			"duration", time.Since(started),
			)
	})
}
```
